IdentityCore/src/oauth2/b2c/MSIDB2COauth2Factory.m

```diff
@@ -0,0 +1,133 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import "MSIDB2COauth2Factory.h"
+#import "MSIDB2CTokenResponse.h"
+#import "MSIDOauth2Factory+Internal.h"
+#import "MSIDAccessToken.h"
+#import "MSIDAuthorityFactory.h"
+#import "MSIDIdToken.h"
+
+@implementation MSIDB2COauth2Factory
+
+#pragma mark - Helpers
+
+- (BOOL)checkResponseClass:(MSIDB2CTokenResponse *)response
+                   context:(id<MSIDRequestContext>)context
+                     error:(NSError **)error
+{
+    if (![response isKindOfClass:[MSIDB2CTokenResponse class]])
+    {
+        if (error)
+        {
+            NSString *errorMessage = [NSString stringWithFormat:@"Wrong token response type passed, which means wrong factory is being used (expected MSIDB2CTokenResponse, passed %@", response.class];
+
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+
+        return NO;
+    }
+
+    return YES;
+}
+
+#pragma mark - Response
+
+- (MSIDB2CTokenResponse *)tokenResponseFromJSON:(NSDictionary *)json
+                                        context:(id<MSIDRequestContext>)context
+                                          error:(NSError **)error
+{
+    return [[MSIDB2CTokenResponse alloc] initWithJSONDictionary:json error:error];
+}
+
+- (MSIDB2CTokenResponse *)tokenResponseFromJSON:(NSDictionary *)json
+                                   refreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)token
+                                        context:(id<MSIDRequestContext>)context
+                                          error:(NSError * __autoreleasing *)error
+{
+    return [[MSIDB2CTokenResponse alloc] initWithJSONDictionary:json refreshToken:token error:error];
+}
+
+- (BOOL)verifyResponse:(MSIDB2CTokenResponse *)response
+               context:(id<MSIDRequestContext>)context
+                 error:(NSError * __autoreleasing *)error
+{
+    if (![self checkResponseClass:response context:context error:error])
+    {
+        return NO;
+    }
+
+    return [super verifyResponse:response context:context error:error];
+}
+
+#pragma mark - Tokens
+
+- (BOOL)fillAccount:(MSIDAccount *)account
+       fromResponse:(MSIDB2CTokenResponse *)response
+      configuration:(MSIDConfiguration *)configuration
+{
+    if (![self checkResponseClass:response context:nil error:nil])
+    {
+        return NO;
+    }
+
+    BOOL result = [super fillAccount:account fromResponse:response configuration:configuration];
```

Comment:
Let's do a separate follow up, I'd like to see this method return an account instead of it touching the passed in one.

Reply:
Sure, added on issue here: https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/issues/269